src/main/java/stats/KMeans.java

```diff
@@ -0,0 +1,162 @@
+package stats;
+
+import stats.utils.Cluster;
+import stats.utils.Point;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Implementation of K-Means Clustering Algorithm.
+ */
+public class KMeans {
+
+    public void init() {
+
+        /**
+         * Create clusters and set random centroids.
+         */
+
+        for (int i = 0; i < k; ++i) {
+            Cluster cluster = new Cluster(i);
+            Point centroid = Point.createRandomPoint(MIN_COORDINATE, MAX_COORDINATE);
+            cluster.setCentroid(centroid);
+            clusters.add(cluster);
+        }
+
+    }
+
+    public void calculate() {
```

Comment:
I think you are looking for contract that looks something like this

```

<DataSetType, SimilarityType, initializerType> void cluster(final DataSetType data, 
                                                                                           final SimilarityType similarity, 
                                                                                           final InitializerType init);
```

Reply:
The algorithm does not need to hold a ref to the dataset...

Reply:
to be honest I don't exactly get what you are saying.
do you believe that there should be parameters inside calculate method?

Reply:
where does the dataset for which you want to perform clustering come from? I can't see it. The ```SimilarityType``` is a type that computes similarities between data point. By the way when I say data point I mean it generically. The ```InitializerType``` provides the initialization for the centroids. Is it clear now?